src/coprocessor/dag/expr/builtin_time.rs

```diff
@@ -453,15 +463,45 @@ impl ScalarFunc {
         ctx: &mut EvalContext,
         row: &'a [Datum],
     ) -> Result<Option<Cow<'a, MyDuration>>> {
-        let d0: Cow<'a, MyDuration> = try_opt!(self.children[0].eval_duration(ctx, row));
-        let d1: Cow<'a, MyDuration> = try_opt!(self.children[1].eval_duration(ctx, row));
-        let diff = match d0.to_nanos().checked_sub(d1.to_nanos()) {
-            Some(result) => result,
-            None => return Err(Error::overflow("DURATION", &format!("({} - {})", &d0, &d1))),
+        let arg0: Cow<'a, MyDuration> = try_opt!(self.children[0].eval_duration(ctx, row));
+        let arg1: Cow<'a, MyDuration> = try_opt!(self.children[1].eval_duration(ctx, row));
+        let overflow = Error::overflow("DURATION", &format!("({} - {})", &arg0, &arg1));
+        let res = match arg0.into_owned().checked_sub(&arg1) {
+            Some(res) => res,
+            None => return Err(overflow),
+        };
+        Ok(Some(Cow::Owned(res)))
+    }
+
+    #[inline]
+    pub fn sub_duration_and_string<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, MyDuration>>> {
+        let arg0: Cow<'a, MyDuration> = try_opt!(self.children[0].eval_duration(ctx, row));
+        let arg1: Cow<'a, [u8]> = try_opt!(self.children[1].eval_string(ctx, row));
+        let s = ::std::str::from_utf8(&arg1)?;
+        let arg1 = match MyDuration::parse(&arg1, Time::parse_fsp(s)) {
+            Ok(arg1) => arg1,
+            Err(_) => return Ok(None),
+        };
+        let overflow = Error::overflow("DURATION", &format!("({} - {})", &arg0, &arg1));
+        let res = match arg0.into_owned().checked_sub(&arg1) {
+            Some(res) => res,
+            None => return Err(overflow),
         };
-        let res = MyDuration::from_nanos(diff, d0.fsp().max(d1.fsp()) as i8)?;
```

Comment:
Here the logic seems been changed. We do not allow to do more than one thing in one PR. For example, if one PR both fixes a bug and create a new feature,  we can not easily cherry-pick the current whole PR into any release version if we want to fix the bug only.

Reply:
Ok, I will open another PR to modify it.